referenceqvm/qvm_stabilizer.py

```diff
@@ -0,0 +1,479 @@
+
+        elif isinstance(instruction, JumpTarget):
+            # Label; pass straight over
+            self.program_counter += 1
+
+    def run(self, pyquil_program, classical_addresses=None, trials=1):
+        """
+        Run pyquil program and return the results
+
+        Loads and checks program if all gates are within the stabilizer set.
+        Then executes program.  If measurements are requested then the measured
+        results are returned
+
+        :param Program pyquil_program: a pyquil Program containing only
+                                       CNOT-H-S-MEASUREMENT operations
+        :param classical_addresses: classical addresses to return
+        :param trials: number of times to repeat the execution of the program
+        :return: list of lists of classical memory after each run
+        """
+        self.load_program(pyquil_program)
+
+        if classical_addresses is None:
+            classical_addresses = get_classical_addresses_from_program(pyquil_program)
+
+        results = []
+        for _ in range(trials):
+            # set up stabilizers
+            self.tableau = self._n_qubit_tableau(self.num_qubits)
+            self.kernel()
+            results.append(list(map(int, self.classical_memory[classical_addresses])))
+
+            # reset qvm
+            self.memory_reset()
+            self.program_counter = 0
+
+        return results
+
+    def density(self, pyquil_program):
+        """
+        Run program and compute the density matrix
+
+        Loads and checks program if all gates are within the stabilizer set.
+        Then executes program and returns the final density matrix for the
+        stabilizer
+
+        :param Program pyquil_program: a pyquil Program containing only
+                                       CNOT-H-S-MEASUREMENT operations
+        :return:
+        """
+        self.load_program(pyquil_program)
+        self.tableau = self._n_qubit_tableau(self.num_qubits)
+        self.kernel()
+        stabilizers = binary_stabilizer_to_pauli_stabilizer(self.stabilizer_tableau())
+        pauli_ops = list(map(lambda x: 0.5 * (sI(0) + x), stabilizers))
+        pauli_ops = reduce(lambda x, y: x * y, pauli_ops)
+        return tensor_up(pauli_ops, self.num_qubits)
+
+    def wavefunction(self, program):
+        """
+        Simulate the program and then project out the final state.
+
+        Return the final state as a wavefunction pyquil object
+
+        :param program: pyQuil program composed of stabilizer operations only
+        :return: pyquil.Wavefunction.wavefunction object.
+        """
+        self.load_program(program)
+        self.tableau = self._n_qubit_tableau(self.num_qubits)
+        self.kernel()
+        stabilizers = binary_stabilizer_to_pauli_stabilizer(self.stabilizer_tableau())
+        stabilizer_state = project_stabilized_state(stabilizers)
+        stabilizer_state = np.array(stabilizer_state.todense())  # todense() returns a matrixlib type
```

Comment:
`stabalizer_state.toarray()` does this in one fell swoop

Reply:
👍